site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+    /** @property @var \DateTime The date that grades will be released to students */
+    protected $grade_released_date = null;
+    /** @property @var \DateTime The date after which only instructors may change grades (aka when grades are 'due') */
+    protected $grade_locked_date = null;
+    /** @property @var int The minimum user group that can grade this gradeable (1=instructor) */
+    protected $min_grading_group = 1;
+    /** @property @var string The syllabus classification of this gradeable */
+    protected $syllabus_bucket = "Homework";
+
+    /** @property @var Component[] An array of all of this gradeable's components */
+    protected $components = array();
+
+    /* Properties exclusive to numeric-text/checkpoint gradeables */
+
+    /** @property @var string The overall ta instructions for grading (numeric-text/checkpoint only) */
+    protected $ta_instructions = "";
+
+    /* Properties exclusive to electronic gradeables */
+
+    /** @property @var string The location of the autograding configuration file */
+    protected $autograding_config_path = "";
+    /** @property @var string[] The object that contains the autograding config data */
+    private $autograding_config = null;
+    /** @property @var bool If the gradeable is using vcs upload (true) or manual upload (false) */
+    protected $vcs = false;
+    /** @property @var string The subdirectory within the VCS repository for this gradeable */
+    protected $vcs_subdirectory = "";
+    /** @property @var bool If the gradeable is a team assignment */
+    protected $team_assignment = false;
+    /** @property @var int The maximum team size (if the gradeable is a team assignment) */
+    protected $team_size_max = 0;
+    /** @property @var \DateTime The deadline for joining teams (if the gradeable is a team assignment) */
+    protected $team_lock_date = null;
+    /** @property @var bool If the gradeable is using any manual grading */
+    protected $ta_grading = false;
+    /** @property @var bool If students can view submissions */
+    protected $student_view = false;
+    /** @property @var bool If students can make submissions */
+    protected $student_submit = false;
+    /** @property @var bool If students can download submitted files */
+    protected $student_download = false;
+    /** @property @var bool If students can view/download any version of the submitted files, or just the active version */
+    protected $student_any_version = false;
+    /** @property @var bool If the gradeable uses peer grading */
+    protected $peer_grading = false;
+    /** @property @var int The number of peers each student will be graded by */
+    protected $peer_grade_set = 0;
+    /** @property @var \DateTime The date students can start making submissions */
+    protected $submission_open_date = null;
+    /** @property @var \DateTime The date, before which all students must make a submissions (or be marked late) */
+    protected $submission_due_date = null;
```

Comment:
Should all the date fields be together? What about team dates? With this many dates it might even be helpful to make a class to store them all in one place